Add runtime asserts to AOTI

yushangdi opened the pull request:
Summary:
Solves https://github.com/pytorch/pytorch/issues/151925

Currently, AOTI only generate runtime asserts for unbacked symints. We should generate asserts for all `_assert_scalar` calls in the input graph.

Also factored out the run time assertion logic to a separate function.

        We need to generate runtime asserts directly in Inductor instead
        of just re-using the asserts from input graphs becase we reuse the
        same ShapeEnv as before. In particular, on subsequent graph passes,
        we would immediately turn all of these assertions into noops,
        because when we evaluated their expressions, we would see that
        because we had a deferred runtime assert in the ShapeEnv, we
        know "oh, of course this expression is True" already.
        One example is below:
```
        class Model(torch.nn.Module):
            def forward(self, a, b, c):
                nz = torch.nonzero(a)
                ones = a.new_ones([nz.size(0), b.size(0)])
                torch._check(ones.size(0) >= 1)
                equals = torch.add(ones, c)
                return equals
        torch._dynamo.mark_dynamic(c, 0)
```
        When we re-use the ShapeEnv in Inductor lowering, the check that checks
        a and nonzero have the same shape would be evaluted to True after we resolve
        unbacked bindings using the ShapeEnv.
        See test_unbacked_equals_input_size_runtime_assertion in test_aot_inductor.
        
        
        In addition to the Inductor generated runtime asserts, we also
        need the runtime asserts from the input graph, because some derived
        runtime asserts are not generated in Inductor. One example is
        below:
```
        class Model(torch.nn.Module):
            def forward(self, x):
                y = x.reshape(100, -1).clone()
                y = y + 1
                return y
        
        dynamic_shapes = {
            "x": {0: torch.export.Dim.DYNAMIC},
        }
        x.shape[0] needs to be a multiple of 100.
```
        See test_aoti_runtime_asserts_backed_symint in test_aot_inductor.


Example:

```
    def forward(self):
        arg0_1: "f32[s35]";

        arg0_1, = fx_pytree.tree_flatten_spec([], self._in_spec)
         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:11 in forward, code: y = x.reshape(100, -1).clone()
        sym_size_int: "Sym(s35)" = torch.ops.aten.sym_size.int(arg0_1, 0)

         #
        mod: "Sym(Mod(s35, 100))" = sym_size_int % 100;  sym_size_int = None
        eq_2: "Sym(Eq(Mod(s35, 100), 0))" = mod == 0;  mod = None
        _assert_scalar = torch.ops.aten._assert_scalar.default(eq_2, "Runtime assertion failed for expression Eq(Mod(s35, 100), 0) on node 'eq'");  eq_2 = _assert_scalar = None

         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:11 in forward, code: y = x.reshape(100, -1).clone()
        view: "f32[100, (s35//100)]" = torch.ops.aten.reshape.default(arg0_1, [100, -1]);  arg0_1 = None
        clone: "f32[100, (s35//100)]" = torch.ops.aten.clone.default(view);  view = None

         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:12 in forward, code: y = y + 1
        add_6: "f32[100, 1]" = torch.ops.aten.add.Tensor(clone, 1);  clone = None
        return (add_6,)
```

Generated cpp code:

```
    auto inputs = steal_from_raw_handles_to_raii_handles(input_handles, 1);
    auto arg0_1 = std::move(inputs[0]);
    auto arg0_1_size = arg0_1.sizes();
    int64_t s35 = arg0_1_size[0];
    inputs.clear();
    auto& kernels = static_cast<AOTInductorModelKernels&>(*this->kernels_.get());
    if (!((s35 % 100L) == 0L)) { throw std::runtime_error("Expected Eq(Mod(s35, 100), 0) to be True but received " + std::to_string(s35)); }
```

Test Plan:
```
buck run fbcode//mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r aoti_runtime_asserts_backed_symint
```

Differential Revision: D73596786




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang commented:
I can't tell if I am being too careful. Do we want to roll this out to everything?

henrylhtsang commented:
> > I can't tell if I am being too careful. Do we want to roll this out to everything at once?
> 
> @henrylhtsang Is the worry that having more assert will slow down performance of some models?
> 
> If you set TORCHINDUCTOR_SCALAR_ASSERTS=0, then no scalar asserts will be generated (it's default to 1).
> 
> Not sure what's a safer roll out plan, but I can implement it if you have any suggestion!

I wasn't too worried about perf I guess. I was just worrying this would cause some actual AOTI runs to fail immediately. 

We can guard it with a config under aot_inductor? We can set on by default for OSS, then roll out in fbcode, remove the config and the guard in 1 month.

cc @chenyang78 and @desertfire for thoughts, maybe I am just too paranoid

yushangdi commented:
> > > I can't tell if I am being too careful. Do we want to roll this out to everything at once?
> > 
> > 
> > @henrylhtsang Is the worry that having more assert will slow down performance of some models?
> > If you set TORCHINDUCTOR_SCALAR_ASSERTS=0, then no scalar asserts will be generated (it's default to 1).
> > Not sure what's a safer roll out plan, but I can implement it if you have any suggestion!
> 
> I wasn't too worried about perf I guess. I was just worrying this would cause some actual AOTI runs to fail immediately.
> 
> We can guard it with a config under aot_inductor? We can set on by default for OSS, then roll out in fbcode, remove the config and the guard in 1 month.
> 
> cc @chenyang78 and @desertfire for thoughts, maybe I am just too paranoid

sure sounds good to me! Being more cautious is never wrong. Would it be better to use the justknob for the fbcode guard?

jingsh commented:
> > > > I can't tell if I am being too careful. Do we want to roll this out to everything at once?
> > > 
> > > 
> > > @henrylhtsang Is the worry that having more assert will slow down performance of some models?
> > > If you set TORCHINDUCTOR_SCALAR_ASSERTS=0, then no scalar asserts will be generated (it's default to 1).
> > > Not sure what's a safer roll out plan, but I can implement it if you have any suggestion!
> > 
> > 
> > I wasn't too worried about perf I guess. I was just worrying this would cause some actual AOTI runs to fail immediately.
> > We can guard it with a config under aot_inductor? We can set on by default for OSS, then roll out in fbcode, remove the config and the guard in 1 month.
> > cc @chenyang78 and @desertfire for thoughts, maybe I am just too paranoid
> 
> sure sounds good to me! Being more cautious is never wrong. Would it be better to use the justknob for the fbcode guard?

guard would be great!

henrylhtsang commented:
> > > > I can't tell if I am being too careful. Do we want to roll this out to everything at once?
> > > 
> > > 
> > > @henrylhtsang Is the worry that having more assert will slow down performance of some models?
> > > If you set TORCHINDUCTOR_SCALAR_ASSERTS=0, then no scalar asserts will be generated (it's default to 1).
> > > Not sure what's a safer roll out plan, but I can implement it if you have any suggestion!
> > 
> > 
> > I wasn't too worried about perf I guess. I was just worrying this would cause some actual AOTI runs to fail immediately.
> > We can guard it with a config under aot_inductor? We can set on by default for OSS, then roll out in fbcode, remove the config and the guard in 1 month.
> > cc @chenyang78 and @desertfire for thoughts, maybe I am just too paranoid
> 
> sure sounds good to me! Being more cautious is never wrong. Would it be better to use the justknob for the fbcode guard?

just a flag should be enough in my opinion, but either is fine

yushangdi commented:
@henrylhtsang  I added both a env var `"TORCHINDUCTOR_SCALAR_ASSERTS_FULL"` and a justknob now. You'll need to see the internal Diff for the justknob and the flag. The env var overrides the justknob.

henrylhtsang reviewed (approved):
left a comment

yushangdi commented:
there're some errors because the extra asserts incremented the buf name index, so the string match failed. Will fix.

yushangdi commented:
also I think we might need https://github.com/pytorch/pytorch/pull/151919 to fix some of the CI errors related to unbacked symints in the input shape. will try again after https://github.com/pytorch/pytorch/pull/151919  lands

yushangdi committed:
Add runtime asserts to AOTI (#152125)

Summary:

Solves https://github.com/pytorch/pytorch/issues/151925

Currently, AOTI only generate runtime asserts for unbacked symints. We should generate asserts for all `_assert_scalar` calls in the input graph.

It's guarded behind the `aoti_full_runtime_assert` justknob.


Example:

```
    def forward(self):
        arg0_1: "f32[s35]";

        arg0_1, = fx_pytree.tree_flatten_spec([], self._in_spec)
         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:11 in forward, code: y = x.reshape(100, -1).clone()
        sym_size_int: "Sym(s35)" = torch.ops.aten.sym_size.int(arg0_1, 0)

         #
        mod: "Sym(Mod(s35, 100))" = sym_size_int % 100;  sym_size_int = None
        eq_2: "Sym(Eq(Mod(s35, 100), 0))" = mod == 0;  mod = None
        _assert_scalar = torch.ops.aten._assert_scalar.default(eq_2, "Runtime assertion failed for expression Eq(Mod(s35, 100), 0) on node 'eq'");  eq_2 = _assert_scalar = None

         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:11 in forward, code: y = x.reshape(100, -1).clone()
        view: "f32[100, (s35//100)]" = torch.ops.aten.reshape.default(arg0_1, [100, -1]);  arg0_1 = None
        clone: "f32[100, (s35//100)]" = torch.ops.aten.clone.default(view);  view = None

         # File: /data/users/shangdiy/fbsource/buck-out/v2/gen/fbcode/73a672eb896e7996/scripts/shangdiy/__pt__/pt#link-tree/scripts/shangdiy/pt.py:12 in forward, code: y = y + 1
        add_6: "f32[100, 1]" = torch.ops.aten.add.Tensor(clone, 1);  clone = None
        return (add_6,)
```

Generated cpp code:

```
    auto inputs = steal_from_raw_handles_to_raii_handles(input_handles, 1);
    auto arg0_1 = std::move(inputs[0]);
    auto arg0_1_size = arg0_1.sizes();
    int64_t s35 = arg0_1_size[0];
    inputs.clear();
    auto& kernels = static_cast<AOTInductorModelKernels&>(*this->kernels_.get());
    if (!((s35 % 100L) == 0L)) { throw std::runtime_error("Expected Eq(Mod(s35, 100), 0) to be True but received " + std::to_string(s35)); }
```


#buildmore

Test Plan:
```
buck run fbcode//mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r aoti_runtime_asserts_backed_symint

buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchinductor_dynamic_shapes -- -r test_unbacked_floordiv_simplify

TORCHINDUCTOR_SCALAR_ASSERTS_FULL=1 buck run fbcode//mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r test_sym_i64_input_codegen_cuda

TORCHINDUCTOR_SCALAR_ASSERTS_FULL=1  buck run fbcode//mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r  test_unbacked_equals_input_size
```

Reviewed By: henrylhtsang

Differential Revision: D73596786

yushangdi commented:
@henrylhtsang Finally the PR is ready to land! It's changed quite a bit from the original PR, mainly because I did a little code refactoring, can I get a review again? (I'll need to final review after it lands anyway). Thanks!

yushangdi commented on torch/_inductor/graph.py:
```diff
@@ -1872,38 +1863,79 @@ def make_assert(expr: SympyBoolean, msg: str) -> None:
+        return result
 
-        # Emit code for runtime asserts that can be inserted at this point.
-        for i0 in new_unbacked_defs:
```
refactored into function `create_deferred_runtime_asserts`

henrylhtsang commented:
Thanks a lot for the great work!

laithsakka commented on torch/_utils_internal.py:
```diff
@@ -159,6 +159,10 @@ def export_training_ir_rollout_check() -> bool:
     return True
 
 
+def full_aoti_runtime_assert() -> bool:
```
is this meant to be a config cotnrolled? its always true?

yushangdi replied:
> is this meant to be a config controlled? its always true?

it's always True in OSS, internally, we have an env variable and a justknob to control it. Please see the internal Diff for those.

laithsakka commented on torch/_inductor/graph.py:
```diff
@@ -1872,38 +1863,79 @@ def make_assert(expr: SympyBoolean, msg: str) -> None:
+        if (
+            full_aoti_runtime_assert()
+            and n.target == torch.ops.aten._assert_scalar.default
+            and self.aot_mode
```
in general probably eventually we want to add those even for non aot mode.
regular torch.compile should emit those, apparently we can have them if prefer_deferred_runtime_asserts_over_guards.
I am not sure how urgnet or important that is right now.

laithsakka commented on torch/_inductor/graph.py:
```diff
@@ -1872,38 +1863,79 @@ def make_assert(expr: SympyBoolean, msg: str) -> None:
+        ):
+            node_args, _ = self.fetch_args_kwargs_from_env(n)
+            # some assert may have been captured by unbacked symint assertion
+            if node_args[0] != True:  # noqa: E712
```
mmm i do not understand why we skip it if its True?

yushangdi replied:
> mmm i do not understand why we skip it if its True? i mean all of the runtime assertions should be true no? except when reasoning is not smart eough to figure it out can you share an example of such thing where we ignore those.

Do you mean an example where `node_args[0] ` is True or an example where `node_args[0] ` is not True?
`test_unbacked_equals_input_size_runtime_assertion` has an example where we get `True`, and `test_aoti_runtime_asserts_backed_symint` has an example where we get an actual expression.

laithsakka commented on torch/_inductor/graph.py:
```diff
@@ -1872,38 +1863,79 @@ def make_assert(expr: SympyBoolean, msg: str) -> None:
+        if (
+            full_aoti_runtime_assert()
+            and n.target == torch.ops.aten._assert_scalar.default
+            and self.aot_mode
```
thinking out loud, there must be a reason why for unbacked we did not just depend on torch.ops.aten._assert_scalar.default to generate the runtime assert.
the same reason which idk what it is might apply here?

yushangdi replied:
> thinking out loud, there must be a reason why for unbacked we did not just depend on torch.ops.aten._assert_scalar.default to generate the runtime assert. the same reason which idk what it is might apply here?

I added some comments in https://github.com/pytorch/pytorch/pull/153182 explaining why. Ed also explains it in the summary of https://github.com/pytorch/pytorch/pull/124874. 

The problem now is that the inductor generated runtime assertions do not contain all assertions in the exported program. Another potential solution is to figure out why the inductor generated assertions are missing some assertions, and we fix it. For example, `test_aoti_runtime_asserts_backed_symint` has an assertion `size//100 ==0` which is not generated by the inductor, but it's in exported program.

yushangdi commented:
close, relanded in https://github.com/pytorch/pytorch/pull/153182
